lighthouse-core/audits/treemap-data.js

```diff
@@ -0,0 +1,315 @@
+ */
+'use strict';
+
+/**
+ * @fileoverview
+ * Creates treemap data for treemap app.
+ */
+
+const Audit = require('./audit.js');
+const JsBundles = require('../computed/js-bundles.js');
+const UnusedJavaScriptSummary = require('../computed/unused-javascript-summary.js');
+const ModuleDuplication = require('../computed/module-duplication.js');
+const NetworkRecords = require('../computed/network-records.js');
+const ResourceSummary = require('../computed/resource-summary.js');
+
+/**
+ * A collection of root nodes, grouped by type.
+ * @typedef {Record<string, RootNode[]>} TreemapData
+ */
+
+/**
+ * Ex: https://gist.github.com/connorjclark/0ef1099ae994c075e36d65fecb4d26a7
+ * @typedef RootNode
+ * @property {string} name Arbitrary name identifier. Usually a script url.
+ * @property {Node} node
+ */
+
+/**
+ * @typedef Node
+ * @property {string} name Arbitrary name identifier. Usually a path component from a source map.
+ * @property {number} resourceBytes
+ * @property {number=} unusedBytes
+ * @property {string=} duplicate If present, this module is a duplicate. String is normalized source path. See ModuleDuplication.normalizeSource
+ * @property {Node[]=} children
+ */
+
+/**
+ * @typedef {Omit<Node, 'name'|'children'>} SourceData
+ */
+
+class TreemapDataAudit extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'treemap-data',
+      scoreDisplayMode: Audit.SCORING_MODES.INFORMATIVE,
+      title: 'Treemap Data',
+      description: 'Used for treemap app',
+      requiredArtifacts:
+        ['traces', 'devtoolsLogs', 'SourceMaps', 'ScriptElements', 'JsUsage', 'URL'],
+    };
+  }
+
+  /**
+   * Returns a tree data structure where leaf nodes are sources (ie. real files from source tree)
+   * from a source map, and non-leaf nodes are directories. Leaf nodes have data
+   * for bytes, coverage, etc., when available, and non-leaf nodes have the
+   * same data as the sum of all descendant leaf nodes.
+   * @param {string} sourceRoot
+   * @param {Record<string, SourceData>} sourcesData
+   * @return {Node}
+   */
+  static prepareTreemapNodes(sourceRoot, sourcesData) {
+    /**
+     * @param {string} name
+     * @return {Node}
+     */
+    function newNode(name) {
+      return {
+        name,
+        resourceBytes: 0,
+      };
+    }
+
+    /**
+     * Given a slash-delimited path, traverse the Node structure and increment
+     * the data provided for each node in the chain. Creates nodes as needed.
+     * Ex: path/to/file.js will find or create "path" on `node`, increment the data fields,
+     *     and continue with "to", and so on.
+     * @param {string} source
+     * @param {SourceData} data
+     * @param {Node} node
+     */
+    function addNode(source, data, node) {
```

Comment:
`addAllNodesInPath` maybe? feels a little too close to `newNode` for all the work it does.